libp2p-floodsub/src/lib.rs

```diff
@@ -0,0 +1,640 @@
+        });
+    }
+
+    // Internal function that dispatches an `RPC` protobuf struct to all the connected remotes
+    // for which `filter` returns true.
+    fn broadcast<F>(&self, message: rpc_proto::RPC, mut filter: F)
+    where
+        F: FnMut(&FnvHashSet<TopicHash>) -> bool,
+    {
+        let bytes = message
+            .write_to_bytes()
+            .expect("protobuf message is always valid");
+
+        let remote_connections = self.inner.remote_connections.upgradable_read();
+
+        // Number of remotes we dispatched to, for logging purposes.
+        let mut num_dispatched = 0;
+        // Will store the addresses of remotes which we failed to send a message to and which
+        // must be removed from the active connections.
+        let mut failed_to_send: SmallVec<[_; 6]> = SmallVec::new();
+        for (remote_addr, remote) in remote_connections.iter() {
+            if !filter(&remote.subscribed_topics.read()) {
+                continue;
+            }
+
+            num_dispatched += 1;
+            match remote.sender.unbounded_send(bytes.clone().into()) {
+                Ok(_) => (),
+                Err(_) => {
+                    trace!(target: "libp2p-floodsub", "Failed to dispatch message to {} because \
+                                                       channel was closed", remote_addr);
+                    failed_to_send.push(remote_addr.clone());
+                }
+            }
+        }
+
+        // Remove the remotes which we failed to send a message to.
+        if !failed_to_send.is_empty() {
+            // If we fail to upgrade the read lock to a write lock, just ignore `failed_to_send`.
+            if let Ok(mut remote_connections) = remote_connections.try_upgrade() {
+                for failed_to_send in failed_to_send {
+                    remote_connections.remove(&failed_to_send);
+                }
+            }
+        }
+
+        debug!(target: "libp2p-floodsub", "Message queued for {} remotes", num_dispatched);
+    }
+}
+
+/// Implementation of `Stream` that provides messages for the subscribed topics you subscribed to.
+pub struct FloodSubReceiver {
+    inner: mpsc::UnboundedReceiver<Message>,
+}
+
+impl Stream for FloodSubReceiver {
+    type Item = Message;
+    type Error = IoError;
+
+    #[inline]
+    fn poll(&mut self) -> Poll<Option<Self::Item>, Self::Error> {
+        self.inner
+            .poll()
+            .map_err(|_| unreachable!("UnboundedReceiver cannot err"))
+    }
+}
+
+impl fmt::Debug for FloodSubReceiver {
+    fn fmt(&self, fmt: &mut fmt::Formatter) -> fmt::Result {
+        fmt.debug_struct("FloodSubReceiver").finish()
+    }
+}
+
+/// A message received by the floodsub system.
+#[derive(Debug, Clone, PartialEq, Eq, Hash)]
+pub struct Message {
+    /// Remote that sent the message.
+    pub source: Multiaddr,
+
+    /// Content of the message. Its meaning is out of scope of this library.
+    pub data: Vec<u8>,
```

Comment:
Maybe `Bytes`?

Reply:
I went with `Vec<u8>` because the `protobuf` library needs a `Vec<u8>` for its data.